os/kernel/sched/sched_nanosleep.c

```diff
@@ -0,0 +1,243 @@
+		 */
+
+#ifdef CONFIG_HAVE_LONG_LONG
+		uint64_t waitticks64 = ((uint64_t) timeout->tv_sec * NSEC_PER_SEC + (uint64_t) timeout->tv_nsec + NSEC_PER_TICK - 1) / NSEC_PER_TICK;
+		DEBUGASSERT(waitticks64 <= UINT32_MAX);
+		waitticks = (uint32_t) waitticks64;
+#else
+		uint32_t waitmsec;
+
+		DEBUGASSERT(timeout->tv_sec < UINT32_MAX / MSEC_PER_SEC);
+		waitmsec = timeout->tv_sec * MSEC_PER_SEC + (timeout->tv_nsec + NSEC_PER_MSEC - 1) / NSEC_PER_MSEC;
+		waitticks = MSEC2TICK(waitmsec);
+#endif
+
+		/* Create a watchdog */
+
+		rtcb->waitdog = wd_create();
+		DEBUGASSERT(rtcb->waitdog);
+
+		if (rtcb->waitdog) {
+			/* This little bit of nonsense is necessary for some
+			 * processors where sizeof(pointer) < sizeof(uint32_t).
+			 * see wdog.h.
+			 */
+
+			wdparm_t wdparm;
+			wdparm.pvarg = (FAR void *)rtcb;
+
+			/* Start the watchdog */
+
+			wd_start(rtcb->waitdog, waitticks, (wdentry_t) sleep_timeout, 1, wdparm.dwarg);
+
+			/* Now wait for the watchdog */
+
+			up_block_task(rtcb, TSTATE_SLEEP_NOSIGNAL);
+
+			/* We no longer need the watchdog */
+
+			wd_delete(rtcb->waitdog);
+			rtcb->waitdog = NULL;
+
+			ret = 0;
+		}
+
+	}
+
+	irqrestore(saved_state);
+	sched_unlock();
+	leave_cancellation_point();
+	return ret;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+/****************************************************************************
+ * Name: nanosleep
+ *
+ * Description:
+ *   The nanosleep() function causes the current thread to be suspended from
+ *   execution until the time interval specified by the rqtp argument has elapsed
+ *   The suspension time may be longer than requested because the
+ *   argument value is rounded up to an integer multiple of the sleep
+ *   resolution or because of the scheduling of other activity by the system. But,
+ *   the suspension time will not be less than the time specified by rqtp, as
+ *   measured by the system clock, CLOCK_REALTIME.
+ *
+ * Parameters:
+ *   rqtp - The amount of time to be suspended from execution.
+ *   rmtp - Ignored
+ *	    This parameter is kept here to maintain compatibility with sig_nanosleep() API.
+ *	    Since signals are disabled, the sleep operation cannot be interrupted.
+ *	    So, we are sure that we will always sleep for at least the requested period.
+ *
+ * Returned Value:
+ *   If the nanosleep() function returns because the requested time has
+ *   elapsed, its return value is zero.
+ *
+ *   If nanosleep() fails, it returns a value of -1 and sets errno to
+ *   indicate the error. The nanosleep() function will fail if:
+ *
+ *     EINVAL - The rqtp argument specified a nanosecond value less than
+ *       zero or greater than or equal to 1000 million.
+ *
+ ****************************************************************************/
+
+int nanosleep(FAR const struct timespec *rqtp, FAR struct timespec *rmtp)
```

Comment:
nanosleep API is called by sleep and usleep from the c library. 
So, instead of signal based implementation of nanosleep, I have just created a similar implementation which does not make use of signal.